src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -31,19 +30,11 @@ def _insert_installed(installed, others):
     This iterator is used when the resolver prefers to upgrade an
     already-installed package. Candidates from index are returned in their
     normal ordering, except replaced when the version is already installed.
-
-    Since candidates from index are already sorted by reverse version order,
-    `sorted()` here would keep the ordering mostly intact, only shuffling the
-    already-installed candidate into the correct position. We put the already-
-    installed candidate in front of those from the index, so it's put in front
-    after sorting due to Python sorting's stableness guarentee.
     """
-    candidates = sorted(
-        itertools.chain([installed], others),
-        key=operator.attrgetter("version"),
-        reverse=True,
-    )
-    return iter(candidates)
+    for candidate in others:
+        if candidate.version == installed.version:
+            candidate = installed
+        yield candidate
```

Comment:
Hmmm...

The newer variant isn't strictly equivalent -- instead of 20.3 (installed) and 20.3 (on PyPI), which I imagine will cause behavior changes for install-most-recent-package --force-reinstall?

Worthwhile to add a test for that?

Reply:
The new resolver would set the `installed` to `None` for `--force-reinstall`, and never enter this function. I think there’s already a test ensuring `--force-reinstall` works.

Reply:
Hmm... well, it's still failing 2 new_resolver tests.

https://dev.azure.com/pypa/pip/_build/results?buildId=30461&view=logs&j=88d9dc86-3d79-5e8a-7a2d-b92d46a16d40&t=6bf4b43e-ad63-576f-33cc-d8124dfcf590&l=831

```
FAILED tests/functional/test_new_resolver.py::test_new_resolver_prefers_installed_in_upgrade_if_latest
FAILED tests/functional/test_new_resolver.py::test_new_resolver_skip_inconsistent_metadata
```